File: python/src/equities_bench/env.py
from __future__ import annotations

import os
import platform
import sys
from dataclasses import dataclass


@dataclass
class EnvInfo:
    os: str
    arch: str
    cpus: int
    mem_bytes: int
    os_pretty: str
    cpu_model: str
    runtime: str


def _read(path: str) -> str:
    try:
        with open(path, encoding="utf-8") as f:
            return f.read()
    except OSError:
        return ""


def _find_prefixed(text: str, key: str) -> str | None:
    want = key.lower()
    for line in text.splitlines():
        if ":" not in line:
            continue
        k, v = line.split(":", 1)
        if k.strip().lower() == want:
            return v.strip()
    return None


def collect() -> EnvInfo:
    mem = 0
    meminfo = _read("/proc/meminfo")
    for line in meminfo.splitlines():
        if line.startswith("MemTotal:"):
            digits = "".join(ch for ch in line if ch.isdigit())
            if digits:
                mem = int(digits) * 1024
            break
    osrel = _read("/etc/os-release")
    pretty = _find_prefixed(osrel, "PRETTY_NAME") or f"{platform.system()} {platform.release()}"
    if pretty.startswith('"') and pretty.endswith('"'):
        pretty = pretty[1:-1]
    cpuinfo = _read("/proc/cpuinfo")
    cpu = _find_prefixed(cpuinfo, "model name") or _find_prefixed(cpuinfo, "Hardware")
    if not cpu:
        impl = _find_prefixed(cpuinfo, "CPU implementer")
        part = _find_prefixed(cpuinfo, "CPU part")
        if impl or part:
            cpu = f"implementer {impl or '?'} part {part or '?'}"
    if not cpu:
        cpu = platform.machine()
    return EnvInfo(
        os=sys.platform if sys.platform != "linux" else "linux",
        arch=platform.machine(),
        cpus=os.cpu_count() or 1,
        mem_bytes=mem,
        os_pretty=pretty,
        cpu_model=cpu,
        runtime=f"CPython {sys.version.split()[0]}",
    )

File: rust/src/env.rs
use std::fs;

pub struct EnvInfo {
    pub os: String,
    pub arch: String,
    pub cpus: u32,
    pub mem_bytes: u64,
    pub os_pretty: String,
    pub cpu_model: String,
    pub runtime: String,
}

impl EnvInfo {
    pub fn collect() -> Self {
        let cpus = std::thread::available_parallelism()
            .map(|n| n.get() as u32)
            .unwrap_or(1);
        Self {
            os: std::env::consts::OS.to_string(),
            arch: std::env::consts::ARCH.to_string(),
            cpus,
            mem_bytes: read_mem_total(),
            os_pretty: read_pretty_name(),
            cpu_model: read_cpu_model(),
            runtime: env!("RUSTC_VERSION").to_string(),
        }
    }
}

fn read_mem_total() -> u64 {
    let Ok(text) = fs::read_to_string("/proc/meminfo") else {
        return 0;
    };
    for line in text.lines() {
        if let Some(rest) = line.strip_prefix("MemTotal:") {
            let num: String = rest.chars().filter(|c| c.is_ascii_digit()).collect();
            if let Ok(kb) = num.parse::<u64>() {
                return kb.saturating_mul(1024);
            }
        }
    }
    0
}

fn read_pretty_name() -> String {
    let Ok(text) = fs::read_to_string("/etc/os-release") else {
        return format!("{} {}", std::env::consts::OS, std::env::consts::ARCH);
    };
    for line in text.lines() {
        if let Some(v) = line.strip_prefix("PRETTY_NAME=") {
            return v.trim().trim_matches('"').to_string();
        }
    }
    std::env::consts::OS.to_string()
}

fn read_cpu_model() -> String {
    let Ok(text) = fs::read_to_string("/proc/cpuinfo") else {
        return std::env::consts::ARCH.to_string();
    };
    let lines: Vec<&str> = text.lines().collect();
    if let Some(v) = find_prefixed(&lines, "model name") {
        return v;
    }
    if let Some(v) = find_prefixed(&lines, "Hardware") {
        return v;
    }
    let impl_ = find_prefixed(&lines, "CPU implementer");
    let part = find_prefixed(&lines, "CPU part");
    if impl_.is_some() || part.is_some() {
        return format!(
            "implementer {} part {}",
            impl_.unwrap_or_else(|| "?".into()),
            part.unwrap_or_else(|| "?".into())
        );
    }
    std::env::consts::ARCH.to_string()
}

fn find_prefixed(lines: &[&str], key: &str) -> Option<String> {
    let key = key.to_ascii_lowercase();
    for line in lines {
        let Some((k, v)) = line.split_once(':') else {
            continue;
        };
        if k.trim().eq_ignore_ascii_case(&key) {
            return Some(v.trim().to_string());
        }
    }
    None
}
